framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,726 @@
+
+        AMethod method = clazz.methods.vivify(JVMNames.getJVMMethodName(methodTree));
+        // Method return type
+        AnnotatedTypeMirror lhsATM = atf.getAnnotatedType(methodTree).getReturnType();
+        // Type of the expression returned
+        AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(retNode.getTree().getExpression());
+        updateAnnotationSetInScene(
+                method.returnType, atf, jaifPath, rhsATM, lhsATM,
+                DefaultLocation.RETURNS);
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene.
+     *   <ul>
+     *     <li>If there was no previous annotation for that location, then the
+     *      updated set will be the annotations in newATM.</li>
+     *     <li>If there was a previous annotation, the updated set will be the
+     *      LUB between the previous annotation and newATM.</li>
+     *   </ul>
+     * <p>
+     * @param type ATypeElement of the Scene which will be modified.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used.
+     * @param jaifPath used to identify a Scene.
+     * @param newATM the type containing annotations to be added (or "lubbed")
+     * to the Scene.
+     * @param defLoc the location where the annotation will be added.
+     */
+    private static void updateAnnotationSetInScene(ATypeElement type,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM, AnnotatedTypeMirror oldATM,
+            DefaultLocation defLoc) {
+        AnnotatedTypeMirror curAnnosATM = typeElementToATM(
+                type, atf, newATM.getUnderlyingType());
+        if (curAnnosATM.getAnnotations().size() == newATM.getAnnotations().size()
+                && !newATM.getAnnotations().isEmpty()) {
+            newATM = AnnotatedTypes.leastUpperBound(
+                    atf.getProcessingEnv(), atf, newATM, curAnnosATM);
+        }
+        updateTypeElementFromATM(newATM, oldATM, atf, type, 1, defLoc);
+        modifiedScenes.add(jaifPath);
+    }
+
+    /**
+     * Returns the AClass in an AScene, given a className and a jaifPath.
+     */
+    private static AClass getAClass(String className, String jaifPath) {
+        // Possibly reads .jaif file to obtain a Scene.
+        AScene scene = getScene(jaifPath);
+        return scene.classes.vivify(className);
+    }
+
+    /**
+     * Returns true if am should not be inserted in source code. This happens
+     * when am cannot be inserted in source code or is the default for the
+     * location passed as argument.
+     * <p>
+     * For example, {@link org.checkerframework.common.value.qual.BottomVal}.
+     * Returns false otherwise.
+     * <p>
+     * Invisible qualifiers, which are annotations that contain the
+     * {@link org.checkerframework.framework.qual.InvisibleQualifier}
+     * meta-annotation, also return true.
+     */
+    private static boolean shouldIgnore(AnnotationMirror am,
```

Comment:
I am nervous about having logic from GenericATF.createQualifierDefaults copied to a second place.
This will lead to debugging and maintenance problems. On the other hand, the qualifier defaults
infrastructure is not currently made available to the rest of the framework. At the very least,
I recommend adding a comment here to compare against GenericATF.createQualifierDefaults if this
code changes, and a similar comment at the top of createQualifierDefaults to look here if anything
changed there, as well as a TODO to merge the logic in both locations.